reconcile/utils/acs_api.py

```diff
@@ -0,0 +1,328 @@
+from typing import (
+    Any,
+    Optional,
+    Self,
+)
+
+import requests
+from pydantic import BaseModel
+
+
+class Role(BaseModel):
+    name: str
+    permission_set_id: str
+    access_scope_id: str
+    description: str
+    system_default: bool
+
+    def __init__(self, api_data: Any) -> None:
+        # attributes defined within stackrox(ACS) API for GET /v1/roles
+        check_len_attributes(
+            ["name", "permissionSetId", "accessScopeId"],
+            api_data,
+        )
+
+        # traits is populated for system default resources and contains `origin: DEFAULT`
+        # this attribute is used to ignore such resources from reconciliation
+        traits = api_data.get("traits")
+        is_default = traits is not None and traits.get("origin") == "DEFAULT"
+
+        super().__init__(
+            name=api_data["name"],
+            permission_set_id=api_data["permissionSetId"],
+            access_scope_id=api_data["accessScopeId"],
+            description=api_data.get("description", ""),
+            system_default=is_default,
+        )
+
+
+class Group(BaseModel):
+    id: str
+    role_name: str
+    auth_provider_id: str
+    key: str
+    value: str
+
+    def __init__(self, api_data: Any) -> None:
+        # attributes defined within stackrox(ACS) API for GET /v1/groups
+        check_len_attributes(["roleName", "props"], api_data)
+        try:
+            check_len_attributes(
+                ["id", "authProviderId", "key", "value"], api_data["props"]
+            )
+        except ValueError as e:
+            # it is valid for the default None group to contain empty key/value
+            if api_data["roleName"] != "None":
+                raise e
+
+        super().__init__(
+            role_name=api_data["roleName"],
+            id=api_data["props"]["id"],
+            auth_provider_id=api_data["props"]["authProviderId"],
+            key=api_data["props"]["key"],
+            value=api_data["props"]["value"],
+        )
+
+
+class AccessScope(BaseModel):
+    id: str
+    name: str
+    description: str
+    clusters: list[str]
+    namespaces: list[dict[str, str]]
+
+    def __init__(self, api_data: Any) -> None:
+        # attributes defined within stackrox(ACS) API for GET /v1/simpleaccessscopes/{id}
+        unrestricted = False
+        try:
+            check_len_attributes(["id", "name", "rules"], api_data)
+        except ValueError as e:
+            # it is valid for the default Unrestricted access scope to have null `rules`
+            if api_data.get("name") != "Unrestricted":
+                raise e
+            unrestricted = True
```

Comment:
Similar here, skip check when it's `Unrestricted`. What's the relationship between `name` is `Unrestricted` and failed check? When `name` is `Unrestricted` but check pass, do you expect `unrestricted` to be `True` or `False`?

If you just want to do a nullable access, you can use
`api_data.get("rules", {}).get("includedClusters", [])`

Reply:
`Unrestricted` is returned as:
```json
    {
      "id": "ffffffff-ffff-fff4-f5ff-ffffffffffff",
      "name": "Unrestricted",
      "description": "Access to all clusters and namespaces",
      "rules": null,
      "traits": {
        "mutabilityMode": "ALLOW_MUTATE",
        "visibility": "VISIBLE",
        "origin": "DEFAULT"
      }
    }
```

whereas all other access scopes contain following attributes within `rules`:
```json
{
      "id": "1b10220a-5fbc-4ad8-afcd-cd9ea3f828fe",
      "name": "ns-exclusives",
      "description": "example of assigning permission at ns scope",
      "rules": {
        "includedClusters": [],
        "includedNamespaces": [
          {
            "clusterName": "some-cluster",
            "namespaceName": "some-ns"
          }
        ],
        "clusterLabelSelectors": [],
        "namespaceLabelSelectors": []
      },
      "traits": null
    }
```
`When name is Unrestricted but check pass` given the api response behavior above, this should always fail when attempting to check `rules`

Reply:
control flow update [here](https://github.com/app-sre/qontract-reconcile/pull/3899/commits/46e19b701c0a7159d6e36acec751026c60c589ce)

Reply:
to avoid mutate `unrestricted`, can be like this:

```python
        unrestricted = api_data["name"] == "Unrestricted"
        if not unrestricted:
            check_len_attributes(["rules"], api_data)
```

Reply:
[update](https://github.com/app-sre/qontract-reconcile/pull/3899/commits/ae91b7eff0e453ef0861bf9a84ea2b3e0c9d1586)